// @fanoutio/grip/build/main.js
"use strict";
// Flatten and export
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
// Classes
var Publisher_1 = require("./engine/Publisher");
Object.defineProperty(exports, "Publisher", { enumerable: true, get: function () { return Publisher_1.default; } });
var GripInstruct_1 = require("./data/GripInstruct");
Object.defineProperty(exports, "GripInstruct", { enumerable: true, get: function () { return GripInstruct_1.default; } });
exports.Auth = __importStar(require("./auth/index"));
var Format_1 = require("./data/Format");
Object.defineProperty(exports, "Format", { enumerable: true, get: function () { return Format_1.default; } });
var Item_1 = require("./data/Item");
Object.defineProperty(exports, "Item", { enumerable: true, get: function () { return Item_1.default; } });
var PublisherClient_1 = require("./engine/PublisherClient");
Object.defineProperty(exports, "PublisherClient", { enumerable: true, get: function () { return PublisherClient_1.default; } });
var Channel_1 = require("./data/Channel");
Object.defineProperty(exports, "Channel", { enumerable: true, get: function () { return Channel_1.default; } });
var Response_1 = require("./data/Response");
Object.defineProperty(exports, "Response", { enumerable: true, get: function () { return Response_1.default; } });
var HttpStreamFormat_1 = require("./data/http/HttpStreamFormat");
Object.defineProperty(exports, "HttpStreamFormat", { enumerable: true, get: function () { return HttpStreamFormat_1.default; } });
var HttpResponseFormat_1 = require("./data/http/HttpResponseFormat");
Object.defineProperty(exports, "HttpResponseFormat", { enumerable: true, get: function () { return HttpResponseFormat_1.default; } });
var WebSocketContext_1 = require("./data/websocket/WebSocketContext");
Object.defineProperty(exports, "WebSocketContext", { enumerable: true, get: function () { return WebSocketContext_1.default; } });
var WebSocketEvent_1 = require("./data/websocket/WebSocketEvent");
Object.defineProperty(exports, "WebSocketEvent", { enumerable: true, get: function () { return WebSocketEvent_1.default; } });
var WebSocketMessageFormat_1 = require("./data/websocket/WebSocketMessageFormat");
Object.defineProperty(exports, "WebSocketMessageFormat", { enumerable: true, get: function () { return WebSocketMessageFormat_1.default; } });
var WebSocketDecodeEventException_1 = require("./data/websocket/WebSocketDecodeEventException");
Object.defineProperty(exports, "WebSocketDecodeEventException", { enumerable: true, get: function () { return WebSocketDecodeEventException_1.default; } });
var ConnectionIdMissingException_1 = require("./data/websocket/ConnectionIdMissingException");
Object.defineProperty(exports, "ConnectionIdMissingException", { enumerable: true, get: function () { return ConnectionIdMissingException_1.default; } });
// Utility Functions
var grip_1 = require("./utilities/grip");
Object.defineProperty(exports, "createGripChannelHeader", { enumerable: true, get: function () { return grip_1.createGripChannelHeader; } });
Object.defineProperty(exports, "parseGripUri", { enumerable: true, get: function () { return grip_1.parseGripUri; } });
var webSocketEvents_1 = require("./utilities/webSocketEvents");
Object.defineProperty(exports, "encodeWebSocketEvents", { enumerable: true, get: function () { return webSocketEvents_1.encodeWebSocketEvents; } });
Object.defineProperty(exports, "decodeWebSocketEvents", { enumerable: true, get: function () { return webSocketEvents_1.decodeWebSocketEvents; } });
Object.defineProperty(exports, "createWebSocketControlMessage", { enumerable: true, get: function () { return webSocketEvents_1.createWebSocketControlMessage; } });
var ws_over_http_1 = require("./utilities/ws-over-http");
Object.defineProperty(exports, "isWsOverHttp", { enumerable: true, get: function () { return ws_over_http_1.isWsOverHttp; } });
Object.defineProperty(exports, "getWebSocketContextFromReq", { enumerable: true, get: function () { return ws_over_http_1.getWebSocketContextFromReq; } });
var jwt_1 = require("./utilities/jwt");
Object.defineProperty(exports, "validateSig", { enumerable: true, get: function () { return jwt_1.validateSig; } });
